mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/RealImapConnection.kt

```diff
@@ -445,7 +449,17 @@ internal class RealImapConnection(
                 throw e
             }
 
-            login()
+            try {
```

Comment:
> Detekt error:
> 
> > `mail/protocols/imap/src/main/java/com/fsck/k9/mail/store/imap/RealImapConnection.kt:444:17:` Too many throw statements in the function saslAuthPlainWithLoginFallback. The maximum number of allowed throw statements is 2. [ThrowsCount]
> 
> Is `ThrowsCount` a useful rule? Should we disable it?

This could be solved by adding a new `tryLogin(e: AuthenticationFailedException)` method:

```
private fun tryLogin(e: AuthenticationFailedException) = try {
        login()
    } catch (inner: AuthenticationFailedException) {
        throw inner
    } catch (inner: IOException) {
        Timber.d(inner, "LOGIN fallback failed")
        throw e
    } catch (inner: MessagingException) {
        Timber.d(inner, "LOGIN fallback failed")
        throw e
    }
```

Reply:
Or name it retryLogin()

Reply:
That still exceeds the ThrowsCount value of 2.

Reply:
Hmm, I changed it to the following and it didn't trigger for me. Also tried on a clean build.

```
    private fun saslAuthPlainWithLoginFallback(): List<ImapResponse> {
        return try {
            saslAuthPlain()
        } catch (e: AuthenticationFailedException) {
            if (!isConnected) {
                throw e
            }

            retryLogin(e)
        }
    }

    private fun retryLogin(e: AuthenticationFailedException) = try {
        login()
    } catch (inner: AuthenticationFailedException) {
        throw inner
    } catch (inner: IOException) {
        Timber.d(inner, "LOGIN fallback failed")
        throw e
    } catch (inner: MessagingException) {
        Timber.d(inner, "LOGIN fallback failed")
        throw e
    }
```

If there is still a issue, we might increase the number of allowed throws.

Reply:
Interesting. The `throw` statements are not flagged if an expression body is used for the function. They are if a block body is used (I'd like to avoid using expression bodies for functions that are multiple lines long).

I'm not sure if the `ThrowsCount` rule is useful at all. Since Kotlin doesn't support multi-catch, we might run into this quite a bit. On the other hand, if this was indeed a function where too many `throw` statements would obscure the code flow, I'm sure a couple of other detekt rules dealing with code complexity would have caught this as well.

Reply:
I wouldn't disable it, as limiting throws is a good practise in Kotlin. We could instead increase the allowed count or just add `@Suppress("ThrowsCount")`.